dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.41/V2_41_45__add_section_config_fields.sql

```diff
@@ -0,0 +1,4 @@
+ALTER TABLE section ADD COLUMN IF NOT EXISTS transposesection BOOLEAN DEFAULT FALSE;
+ALTER table section add column if not exists transposedgroups jsonb default '[]'::jsonb;
```

Comment:
@david-mackessy - I am working on this functionality to add extra configuration options to the Section metadata (see PR for more details). For this field `transposedGroups` , we want to be able to configure 1+ `categories` to be transposed/pivoted when rendering the section form on the Frontend.

I modeled it as a soft link with a `json` field. This felt like the simplest thing to do, and could provide some flexibility with this field (i.e. we might want to transpose data elements too in the future). I even thought the four fields maybe can be one json field, to avoid updating the backend as we experiment with more configuration options if possible. This means that we could have dangling categories when they're deleted but in this context (configuring forms), the FE is very permissive and would fall back to showing the categories that exist similar to what would happen if this was cascade deleted in a relationship.

In the codebase, I see this pattern used in few places (but mostly analytics) but also sometimes we're normalising with a proper relationship (i.e. greyFields). So I am wondering what's the right way to go ahead with this?

(ps: this PR is missing other local files, but just want to discuss this specific point first)

cc' @Birkbjo @stian-sandvold

Reply:
Hi @kabaros, using jsonb will provide you with a lot of flexibility. You can update & query the jsonb model as you please without needing further DB changes. Without knowing the requirements or how large the jsonb may become, just try to cover different scenarios so you know you're happy with performance if the model becomes very large.

It sounds like you're looking for some kind of dynamic table effect, speaking with Analytics (Maikel, Giusseppe, Dusan) may provide useful also.

We use jsonb in 2 different ways (AFAIK):
- As a user type (see the `JsonBinaryType` in our codebase and which classes use it and how it's used,`Sharing` and `Translation` are examples)
- As a plain string value (`DataStore` jbPlainValue is an example)

You probably have to choose which one `transposedGroups` should be. I don't have enough experience with these to suggest which would be better or easier to use.

I'll tag @jbee as well to see if he has any suggestions or concerns.

Reply:
Thanks @david-mackessy for the pointers, I added the field as jsonb, and mapped it with `jbObject` (similar to its use in systemSettings and MapView) - this provides the flexibility for the FE to define any JSON payload without having to update the backend.

Reply:
it'd be great if you can help me with solving the sonar issue - not sure why it's happening as the same pattern exists in [MapView](https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-api/src/main/java/org/hisp/dhis/mapping/MapView.java#L197) and a couple other places.

Reply:
I fixed the sonar issue, by adding `transient` to the `displayOptions` per the suggestion from Sonar itself. Initially I thought this would mean that the property is not serialised in the JSON but that doesn't seem to be the case and it works as expected - let me know if that potentially has other implications.

Reply:
I'm not sure why a map is being used. What will the map keys be? the json root fields? If so, it feels a bit messy doing it like this.
I think you could get what you're looking for with string. The `DatastoreEntry` field `jbPlainValue` is another example of using string to store in a jsonb column.
If you're happy with what's there & it satisfies your requirements then 👍

Reply:
the json I am sending looks like

```json
{
"pivotMode": "x",
"pivotedCategory": "XXX"
}
```

is that not suitable to be modeled as a map of strings to strings? We 'd be adding more properties (these are the keys), but they will always be strings keys to strings values. I will look at the `jbbPlainValue` as well though.

Reply:
ok i see. As you mentioned earlier, it looks like you're settling on a middle ground rather than having the full flexibility. But if you know this will only ever be a straight string -> string mapping then fair enough.